// mocker-api/lib/mockerHandle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mockerHandle = mockerHandle;

var _url = _interopRequireDefault(require("url"));

var _bodyParser = _interopRequireDefault(require("body-parser"));

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function mockerHandle(param) {
  const {
    options = {},
    req,
    res,
    next,
    mocker,
    mockerKey
  } = param || {};

  let bodyParserMethd = _bodyParser.default.json({ ...options.bodyParserJSON
  }); // 默认使用json解析

  /**
   * `application/x-www-form-urlencoded; charset=UTF-8` => `application/x-www-form-urlencoded`
   * Issue: https://github.com/jaywcjlove/mocker-api/issues/50
   */


  let contentType = req.get('Content-Type');
  contentType = contentType && contentType.replace(/;.*$/, '');

  if (options.bodyParserConf && options.bodyParserConf[contentType]) {
    // 如果存在options.bodyParserConf配置 {'text/plain': 'text','text/html': 'text'}
    switch (options.bodyParserConf[contentType]) {
      // 获取bodyParser的方法
      case 'raw':
        bodyParserMethd = _bodyParser.default.raw({ ...options.bodyParserRaw
        });
        break;

      case 'text':
        bodyParserMethd = _bodyParser.default.text({ ...options.bodyParserText
        });
        break;

      case 'urlencoded':
        bodyParserMethd = _bodyParser.default.urlencoded({
          extended: false,
          ...options.bodyParserUrlencoded
        });
        break;

      case 'json':
        bodyParserMethd = _bodyParser.default.json({ ...options.bodyParserJSON
        });
      //使用json解析 break;
    }
  } else {
    // 兼容原来的代码,默认解析
    // Compatible with the original code, default parsing
    switch (contentType) {
      case 'text/plain':
        bodyParserMethd = _bodyParser.default.raw({ ...options.bodyParserRaw
        });
        break;

      case 'text/html':
        bodyParserMethd = _bodyParser.default.text({ ...options.bodyParserText
        });
        break;

      case 'application/x-www-form-urlencoded':
        bodyParserMethd = _bodyParser.default.urlencoded({
          extended: false,
          ...options.bodyParserUrlencoded
        });
        break;
    }
  }

  bodyParserMethd(req, res, function () {
    const result = mocker[mockerKey];

    if (typeof result === 'function') {
      const rgxStr = ~mockerKey.indexOf(' ') ? ' ' : '';
      req.params = (0, _utils.pathMatch)({
        sensitive: false,
        strict: false,
        end: false
      })(mockerKey.split(new RegExp(rgxStr))[1])(_url.default.parse(req.url).pathname);
      result(req, res, next);
    } else {
      res.json(result);
    }
  });
}
